atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java

```diff
@@ -71,18 +71,27 @@ private AtlasDbMetrics() {}
         return instrument(metricRegistry, serviceInterface, service, name, instrumentAllMethods());
     }
 
+    public static <T, U extends T> T instrumentWithTaggedMetrics(
+            TaggedMetricRegistry taggedMetrics,
+            Class<T> serviceInterface,
+            U service) {
+        return Instrumentation.builder(serviceInterface, service)
+                .withHandler(new TaggedMetricsInvocationEventHandler(
+                        taggedMetrics,
+                        MetricRegistry.name(serviceInterface)))
+                .withPerformanceTraceLogging()
+                .build();
+    }
+
```

Comment:
This implementation is very nearly identical to the one below - is there a common method that can be extracted?

Also, below you factored out `String name = MetricRegistry.name(serviceInterface);`, but not in this block.

Reply:
can do, not really sure it buys anything though, this is what it looks like:

```java
    public static <T, U extends T> T instrumentWithTaggedMetrics(
            TaggedMetricRegistry taggedMetrics,
            Class<T> serviceInterface,
            U service) {
        return buildInstrumentation(
                serviceInterface,
                service,
                new TaggedMetricsInvocationEventHandler(taggedMetrics, MetricRegistry.name(serviceInterface)));
    }

    public static <T, U extends T> T instrumentWithTaggedMetrics(
            TaggedMetricRegistry taggedMetrics,
            Class<T> serviceInterface,
            U service,
            Function<InvocationContext, Map<String, String>> tagFunction) {
        return buildInstrumentation(
                serviceInterface,
                service,
                new TaggedMetricsInvocationEventHandler(
                        taggedMetrics,
                        MetricRegistry.name(serviceInterface),
                        tagFunction));
    }

    private static <T, U extends T> T buildInstrumentation(
            Class<T> serviceInterface,
            U service,
            TaggedMetricsInvocationEventHandler handler) {
        return Instrumentation.builder(serviceInterface, service)
                .withHandler(handler)
                .withPerformanceTraceLogging()
                .build();
    }
```

The factoring was so it would not kill my line count aha.

Reply:
Yeah this isn't particularly pretty. Don't mind either way, so long as you are consistent (i.e. the factoring thing is either done or not done for both.